tokamak.py

```diff
@@ -0,0 +1,1190 @@
+        eps = 1e-10
+        psi0 = self.psi(R,Z)
+        dpsidr = (self.psi(R + eps, Z) - psi0) / eps
+        dpsidz = (self.psi(R, Z + eps) - psi0) / eps
+        norm = 1. / (dpsidr**2 + dpsidz**2) # Common factor
+        return dpsidr * norm
+
+    @handleMultiLocationArray
+    def f_Z(self, R, Z):
+        """returns the Z component of the vector Grad(psi)/|Grad(psi)|**2."""
+        # Note: The following lines should work, but
+        # give wrong answers. Instead a simple method is used for now.
+        #dpsidR = self.psi_func(R, Z, dx=1, grid=False)
+        #dpsidZ = self.psi_func(R, Z, dy=1, grid=False)
+        #return dpsidZ / np.sqrt(dpsidR**2 + dpsidZ**2)
+        eps = 1e-10
+        psi0 = self.psi(R,Z)
+        dpsidr = (self.psi(R + eps, Z) - psi0) / eps
+        dpsidz = (self.psi(R, Z + eps) - psi0) / eps
+        norm = 1. / (dpsidr**2 + dpsidz**2) # Common factor
+        return dpsidz * norm
+
+    @handleMultiLocationArray
+    def Bp_R(self, R, Z):
+        """returns the R component of the poloidal magnetic field."""
+        return self.psi_func(R, Z, dy=1, grid=False) / R
+
+    @handleMultiLocationArray
+    def Bp_Z(self, R, Z):
+        """returns the Z component of the poloidal magnetic field."""
+        return -self.psi_func(R, Z, dx=1, grid=False) / R
+
+    @handleMultiLocationArray
+    def fpol(self, psi):
+        """poloidal current function, 
+        returns fpol such that B_toroidal = fpol/R"""
+        return self.f_spl(psi * self.f_psi_sign)
+
+    @handleMultiLocationArray
+    def fpolprime(self, psi):
+        """psi-derivative of fpol
+        """
+        return self.fprime_spl(psi * self.f_psi_sign)
+    
+
+def read_geqdsk(filehandle):
+    """
+    Read geqdsk formatted data from a file object, returning
+    a TokamakEquilibrium object
+    """
+
+    from ._geqdsk import read as geq_read
+    
+    data = geq_read(filehandle)
+    
+    # Range of psi normalises psi derivatives
+    psi_boundary = data["sibdry"]
+    psi_axis = data["simagx"]
+
+    # 1D grid on which fpol is defined. Goes from normalised psi 0 to 1
+    psi1D = np.linspace(psi_axis, psi_boundary,
+                        data["nx"], endpoint=True)
+
+    R1D = np.linspace(data["rleft"], data["rleft"] + data["rdim"],
+                      data["nx"], endpoint=True)
+                            
+    Z1D = np.linspace(data["zmid"] - 0.5*data["zdim"],
+                      data["zmid"] + 0.5*data["zdim"],
+                      data["ny"], endpoint=True)
+
+    # Get the wall 
+    if "rlim" in data and "zlim" in data:
+        wall = list(zip(data["rlim"], data["zlim"]))
+    else:
+        wall = []
+    
+    return TokamakEquilibrium(R1D,
+                              Z1D,
+                              data["psi"],  # psi2D
+                              psi1D, 
+                              data["fpol"], # fpol1D
+                              wall=wall)
+
+def example():
```

Comment:
Might be worth adding an argument to pick LSN, USN, CDN or DDN, since functions are there already.